AimTTL PL series power supplies

guuskuiper opened the pull request:
Adds instruments for the AimTTL PL-P series power supplies.

Protocol documentation:
https://resources.aimtti.com/manuals/New_PL+PL-P_Series_Instruction_Manual-Iss18.pdf

List of PL series power supplies:
https://www.aimtti.com/product-category/dc-power-supplies/aim-plseries

bilderbuchi commented:
Neat, thanks! At the previous job, we had QL355P, PLH250-P, CPX400SP, which all had quite similar command set, I think identical to this one. Sadly I don't have access to the hardware anymore, but maybe those could be included at the same time (probably just a question of appropriate ranges)?

bilderbuchi commented on pymeasure/instruments/aimtti/aimttiPL.py:
```diff
@@ -0,0 +1,176 @@
+        map_values=True
+        )
+
+    def enable(self):
+        """ Enables the output."""
+        self.write("OP{ch} 1")
+
+    def disable(self):
+        """ Disables the output."""
+        self.write("OP{ch} 0")
```
We try to[ call this ](https://pymeasure.readthedocs.io/en/latest/dev/adding_instruments/instrument.html#frequent-properties) `output_enabled` (in one property).

bilderbuchi commented on pymeasure/instruments/aimtti/aimttiPL.py:
```diff
@@ -0,0 +1,176 @@
+        Low (500/800mA) range, or High range. 
+        Output must be switched off before changing range.""",
+        validator=strict_discrete_set,
+        values=CURRENT_RANGE,
```
If that is the only place you use `CURRENT_RANGE`, maybe just add it directly here, and not have another class attribute lying around?

bilderbuchi commented on pymeasure/instruments/aimtti/aimttiPL.py:
```diff
@@ -0,0 +1,176 @@
+        kwargs.setdefault('timeout', 5000)
+        super().__init__(adapter, name, **kwargs)
+
+    def enable(self):
+        """ Simultaneously sets all outputs on."""
+        self.write("OPALL 1")
+
+    def disable(self):
+        """ Simultaneously sets all outputs off."""
+        self.write("OPALL 0")
```
Similar to above, maybe `all_outputs_enabled`?

bilderbuchi commented on pymeasure/instruments/aimtti/aimttiPL.py:
```diff
@@ -0,0 +1,176 @@
+
+    """
+
+    chs = Instrument.ChannelCreator(PLChannel, ("1", "2"), voltage_range=[0, 30], current_range=[0, 3])
+    """ Channels of the power supply. The channels are number from right-to-left, starting at 1. 
+    Default values are for the voltage and current range are for the PL303MQD-P."""
```
I would not define channels in the base instrument at all, as they seem to be overridden anyway?

bilderbuchi commented on pymeasure/instruments/aimtti/aimttiPL.py:
```diff
@@ -0,0 +1,176 @@
+
+class PL068P(PLBase):
+
+    chs = Instrument.ChannelCreator(PLChannel, "1", voltage_range=[0, 6], current_range=[0, 8])
```
This has very recently settled down with the release last week that the behaviour was finalised in a different way, in that for instruments with few channels it's clearer/more explicit to add the channels directly: https://pymeasure.readthedocs.io/en/latest/dev/adding_instruments/channels.html#adding-a-channel-with-channelcreator (`ChannelCreator` now does that)
and only use a `MultiChannelCreator` for complex/large configs.

So, here you'd do `ch_1 = Instrument.ChannelCreator(PLChannel, "1", voltage_range=[0, 6], current_range=[0, 8])`

guuskuiper commented:
> Neat, thanks! At a previous job, we had QL355P, PLH250-P, CPX400SP, which all had quite similar command set, I think identical to this one. Sadly I don't have access to the hardware anymore, but maybe those could be included at the same time (probably just a question of appropriate ranges)? I liked the TTi PSUs, nice hardware 😁

Thanks for the feedback. I'll fix the issues next time I'm working in this code.
The other series of AimTTi PSU indeed seem very similar. Although I did spot a few differences. There are some variation in the output range of the PSU's: RANGE / VRANGE / IRANGE. Also not all devices support the "enable all outputs" command (OPALL). I might make a base channel class that contains all shared functionality, and specific ones that add the RANGE / VRANGE / IRANGE. Or do you have any advice how to handle minor differences in supported commands?

The following series might be compatible:

- PL-L (current code, IRANGE)
- PLH-P (no enabled all outputs)
- TSX II (no enabled all outputs)
- QL II (RANGE, and an additional limited AUX channel)
- MX (VRANGE)
- QPX 
- CPX

guuskuiper commented:
I've process all the comments, and tested it again with a PL303MQD-P device. So the code is ready for a final review. 

The other similar AimTTI device series can be added in a upcoming pull request. Main question for that is how to test those other series, since I only have access to the PL303MQD-P device, and what to do with the small differences in the interface?

bmoneke commented:
> The other similar AimTTI device series can be added in a upcoming pull request. Main question for that is how to test those other series, since I only have access to the PL303MQD-P device, and what to do with the small differences in the interface?

Regarding how to address small differences in the interface, I'd say dynamic properties are best. The documentation talks about similar instruments: https://pymeasure.readthedocs.io/en/latest/dev/adding_instruments/properties.html#instruments-with-similar-features

Regarding testing capabilities: Take the manual and test, whether the code behaves as the manual stipulates. You could state in the instrument description, that it is according to the manual, but not tested with an actual device.

bilderbuchi commented:
Awesome! I'll try to get to reviewing as soon as possible.

bilderbuchi reviewed (changes requested):
Awesome, I like how this supports a whole model family with very little code changes! @msmttchr this really highlights the power of the dynamic properties you added.

The code is looking good, with some minor remarks. 

One thing that is still missing is some files to make this mesh into the pymeasure instrument set, and so that we get the documentation rendered: https://pymeasure.readthedocs.io/en/latest/dev/adding_instruments/instrument.html#file-structure

bilderbuchi commented:
> The other similar AimTTI device series can be added in a upcoming pull request.

Sounds good to me.

guuskuiper commented on tests/instruments/aimtti/test_aimttl.py:
```diff
@@ -0,0 +1,128 @@
+         ],
+    ) as inst:
+        inst.ch_1.voltage_setpoint = 60
+        assert inst.ch_1.voltage_setpoint == 60
```
Good point, nothing, so I remove that test.

guuskuiper commented:
> Awesome, I like how this supports a whole model family with very little code changes! @msmttchr this really highlights the power of the dynamic properties you added.
> 
> The code is looking ready, with some minor remarks.
> 
> One thing that is still missing is some files to make this mesh into the pymeasure instrument set, and so that we get the documentation rendered: https://pymeasure.readthedocs.io/en/latest/dev/adding_instruments/instrument.html#file-structure

I've fixed the remarks, and added the documentation. Although I had some issues with the channels missing in each class as created by the Instrument.ChannelCreator. By adding some more Sphinx code in the `aimttiPL.rst`  I did get it somewhat working. Do you know a better way to do this? Or does this need to be solve in the ChannelCreator / Sphinx configuration?

bilderbuchi commented:
There is something wonky with the Sphinx rendering, I'll try to see if I can improve this.

bilderbuchi commented:
I think I fixed the problem, but it seems I'm not allowed to push to your PR branch, so have these commits as a patch instead:
```patch
From 5a4cc91aaade4d88666f3d6c6d0d01d0007e7cff Mon Sep 17 00:00:00 2001
From: Christoph Buchner <bilderbuchi@phononoia.at>
Date: Sat, 23 Sep 2023 17:56:01 +0200
Subject: [PATCH 1/2] Revert "AimTTi try to fix Sphinx"

This reverts commit 42bdf9cedd5de338991f0eb07e188381212481d3.
---
 docs/api/instruments/aimtti/aimttiPL.rst | 6 ------
 1 file changed, 6 deletions(-)

diff --git a/docs/api/instruments/aimtti/aimttiPL.rst b/docs/api/instruments/aimtti/aimttiPL.rst
index 48720d41..88666440 100644
--- a/docs/api/instruments/aimtti/aimttiPL.rst
+++ b/docs/api/instruments/aimtti/aimttiPL.rst
@@ -15,34 +15,28 @@ Aim-TTI PL Series Power Supplies
     :members:
     :undoc-members:
     :show-inheritance:
-    :noindex:
 
 .. autoclass:: pymeasure.instruments.aimtti.aimttiPL.PL155P
     :members:
     :undoc-members:
     :show-inheritance:
-    :noindex:
 
 .. autoclass:: pymeasure.instruments.aimtti.aimttiPL.PL303P
     :members:
     :undoc-members:
     :show-inheritance:
-    :noindex:
 
 .. autoclass:: pymeasure.instruments.aimtti.aimttiPL.PL601P
     :members:
     :undoc-members:
     :show-inheritance:
-    :noindex:
 
 .. autoclass:: pymeasure.instruments.aimtti.aimttiPL.PL303QMDP
     :members:
     :undoc-members:
     :show-inheritance:
-    :noindex:
 
 .. autoclass:: pymeasure.instruments.aimtti.aimttiPL.PL303QMTP
     :members:
     :undoc-members:
     :show-inheritance:
-    :noindex:
-- 
2.34.1

``` 

```patch
From 4869b561808131edd00c83b809e6eaca0f6ee992 Mon Sep 17 00:00:00 2001
From: Christoph Buchner <bilderbuchi@phononoia.at>
Date: Sat, 23 Sep 2023 18:54:32 +0200
Subject: [PATCH 2/2] Improve AimTTI documentation rendering.

---
 docs/api/instruments/aimtti/aimttiPL.rst  | 41 ++++++++++++++++++++++-
 docs/dev/adding_instruments/solutions.rst |  1 +
 pymeasure/instruments/aimtti/aimttiPL.py  |  3 +-
 3 files changed, 43 insertions(+), 2 deletions(-)

diff --git a/docs/api/instruments/aimtti/aimttiPL.rst b/docs/api/instruments/aimtti/aimttiPL.rst
index 88666440..2e8230e2 100644
--- a/docs/api/instruments/aimtti/aimttiPL.rst
+++ b/docs/api/instruments/aimtti/aimttiPL.rst
@@ -1,42 +1,81 @@
+.. _aimtti-landing-page:
+
 #############################################
 Aim-TTI PL Series Power Supplies
 #############################################
 
 .. autoclass:: pymeasure.instruments.aimtti.aimttiPL.PLBase
     :members:
-    :special-members: __init__
     :show-inheritance:
 
 .. autoclass:: pymeasure.instruments.aimtti.aimttiPL.PLChannel
     :members:
     :show-inheritance:
 
+.. For instruments with channels, we explicitly add an autoattribute with no-value
+.. to avoid "= <...ChannelCreator object>" in the rendered documentation
+.. All explicitly documented members are excluded so that there is no duplication,
+.. but we still get docs for all members that are not explicitly enumerated.
+
 .. autoclass:: pymeasure.instruments.aimtti.aimttiPL.PL068P
     :members:
+    :exclude-members: ch_1
     :undoc-members:
     :show-inheritance:
 
+    .. autoattribute:: ch_1
+       :no-value:
+
 .. autoclass:: pymeasure.instruments.aimtti.aimttiPL.PL155P
     :members:
+    :exclude-members: ch_1
     :undoc-members:
     :show-inheritance:
 
+    .. autoattribute:: ch_1
+       :no-value:
+
 .. autoclass:: pymeasure.instruments.aimtti.aimttiPL.PL303P
     :members:
+    :exclude-members: ch_1
     :undoc-members:
     :show-inheritance:
 
+    .. autoattribute:: ch_1
+       :no-value:
+
 .. autoclass:: pymeasure.instruments.aimtti.aimttiPL.PL601P
     :members:
+    :exclude-members: ch_1
     :undoc-members:
     :show-inheritance:
 
+    .. autoattribute:: ch_1
+       :no-value:
+
 .. autoclass:: pymeasure.instruments.aimtti.aimttiPL.PL303QMDP
     :members:
+    :exclude-members: ch_1, ch_2
     :undoc-members:
     :show-inheritance:
 
+    .. autoattribute:: ch_1
+       :no-value:
+
+    .. autoattribute:: ch_2
+       :no-value:
+
 .. autoclass:: pymeasure.instruments.aimtti.aimttiPL.PL303QMTP
     :members:
+    :exclude-members: ch_1, ch_2, ch_3
     :undoc-members:
     :show-inheritance:
+
+    .. autoattribute:: ch_1
+       :no-value:
+
+    .. autoattribute:: ch_2
+       :no-value:
+
+    .. autoattribute:: ch_3
+       :no-value:
diff --git a/docs/dev/adding_instruments/solutions.rst b/docs/dev/adding_instruments/solutions.rst
index e944c31d..1372d504 100644
--- a/docs/dev/adding_instruments/solutions.rst
+++ b/docs/dev/adding_instruments/solutions.rst
@@ -30,3 +30,4 @@ Channels
 - Not all channels have the same features, see :class:`MKS937B <pymeasure.instruments.mksinst.mks937b.MKS937B>`
 - Channel names in the communication (1, 2, 3) differ from front panel (A, B, C), see :class:`AdvantestR624X <pymeasure.instruments.advantest.advantestR624X.AdvantestR624X>`
 - A family of instruments in which a property of the channels is different for different members of the family , see :class:`AnritsuMS464xB <pymeasure.instruments.anritsu.AnritsuMS464xB>`
+- If you want to document the type of your instrument's channels (with a clickable link), check out the source of the :ref:`aimtti-landing-page` page.
diff --git a/pymeasure/instruments/aimtti/aimttiPL.py b/pymeasure/instruments/aimtti/aimttiPL.py
index 1d319b14..6b09abf0 100644
--- a/pymeasure/instruments/aimtti/aimttiPL.py
+++ b/pymeasure/instruments/aimtti/aimttiPL.py
@@ -97,6 +97,8 @@ class PLBase(Instrument):
     PL-series devices:
         https://www.aimtti.com/product-category/dc-power-supplies/aim-plseries
 
+    The default value for the timeout argument is set to 5000ms.
+
     .. code-block:: python
 
         psu = PL303QMDP("ASRL7::INSTR")
@@ -110,7 +112,6 @@ class PLBase(Instrument):
     """
 
     def __init__(self, adapter, name="AimTTI PL", **kwargs):
-        """The default value for the timeout keyword argument is set to 5000ms."""
         kwargs.setdefault("timeout", 5000)
         super().__init__(adapter, name, **kwargs)
 
-- 
2.34.1


```

bilderbuchi commented:
Let's see what CI says with those applied, if there are any problem annotations left. If not, we're good to go IMO.

guuskuiper commented:
@bilderbuchi the patch seems to be a nice solution for the docs. 
But is there something wrong with the CI pipeline since it's not running for that patch commit?

bilderbuchi commented:
There is a merge conflict in the AUTHORS.txt -- merge from pymeasure master and fix it, then push again and CI should run.

bilderbuchi commented:
I suspect we are somehow generating indices twice, which triggers these warnings.
Locally, if I remove all those occurrences of
```
    .. autoattribute:: ch_XX
       :no-value:
```
from aimTTi.rst, I get no warnings, a nicely rendered page, with instrument entries in the sidebar, the PLChannel docs first, and then the PLChannel word is a link as expected.
![image](https://github.com/pymeasure/pymeasure/assets/327442/7d6b89e0-a89d-4f78-b71c-ee57f1bc9435)

Do you see the same thing if you render the docs locally? If yes, can you push that?
If that works, we should probably adapt the the "instruction comment" starting at line 15. :-P
I'm not sure why we get entries for `ch_X`, even if we `exclude-members` them :shrug:.

guuskuiper commented:
> I suspect we are somehow generating indices twice, which triggers these warnings. Locally, if I remove all those occurrences of
> 
> ```
>     .. autoattribute:: ch_XX
>        :no-value:
> ```
> 
> from aimTTi.rst, I get no warnings, a nicely rendered page, with instrument entries in the sidebar, the PLChannel docs first, and then the PLChannel word is a link as expected. ![image](https://user-images.githubusercontent.com/327442/272700892-7d6b89e0-a89d-4f78-b71c-ee57f1bc9435.png)
> 
> Do you see the same thing if you render the docs locally? If yes, can you push that? If that works, we should probably adapt the the "instruction comment" starting at line 15. :-P I'm not sure why we get entries for `ch_X`, even if we `exclude-members` them 🤷.

I'll try to test that locally. Hopefully that works and the pipeline succeeds.

guuskuiper commented:
Nice, now the pipeline succeeded, @bilderbuchi thanks for the help :)

bilderbuchi commented:
Then I think we are done here, thank you!
